### ISSUE TITLE
mdns through thread networks / border routing

### ISSUE THREAD

**Issue by dakhnod** [openthread/ot-br-posix#2578]
I have the following setup:

A <--thread--> B <--Ethernet--> C

All of the nodes have avahi running, with ipv4 enabled for B and C.
B(avahi) is configured as a reflector. It is supposed to bridge avahi between A and C.
A and B have point-to-point interfaces enabled.

My issue is that A can simply not resolve C.local.
Is there any quirks for mdns through thread?
Maybe multicast ff00::fb doesn't work in thread?

Thanks you so much!

**Comment by jwhui** [comment]
Thread supports DNS-based Service Discovering using the [Service Registration Protocol](https://datatracker.ietf.org/doc/html/draft-ietf-dnssd-srp) and unicast DNS. Thread Border Routers implement both a DNS Discovery Proxy and Advertising Proxy to support discovery across the Thread and Ethernet/Wi-Fi segments.

The [Thread Border Router Codelab](https://openthread.io/codelabs/openthread-border-router) may be useful to provide more information.

**Comment by dakhnod** [comment]
Thanks, this is great!
Yet, does thread anyhow stand in the way of mdns? For our setup, it would be great to avoid having an a centralized dns server...

**Comment by jwhui** [comment]
With SRP, the Thread Border Router serves as a DNS an SRP server. Thread devices publish their DNS service names to the SRP server and the Thread Border Router publishes them using mDNS on the Ethernet side. This allows Ethernet/Wi-Fi hosts to use mDNS to discover services published by Thread devices.

**Comment by dakhnod** [comment]
Okay, this sounds useful.
But what about the other side?
Are thread devices able to discover mdns devices from the ethernet network?

**Comment by jwhui** [comment]
Yes, Thread devices use unicast DNS to the Thread Border Router, which then uses mDNS via a Discovery Proxy to discover services on the Ethernet/Wi-Fi side.

**Comment by dakhnod** [comment]
Ah, so the following apply?
1. I don't need to have avahi running on my thread devices
2. I don't need to configure avahi as reflector, since the BR does the mDNS bridging?

Do I still have to request .local addresses?

**Comment by jwhui** [comment]
> 1. I don't need to have avahi running on my thread devices

Correct.

> 2. I don't need to configure avahi as reflector, since the BR does the mDNS bridging?

The BR needs an mDNS service to perform DNS-SD/mDNS queries and responses. ot-br-posix has example integrations with avahi and mDNSResponder.

> Do I still have to request .local addresses?

Thread devices use `default.service.arpa` to discover services within the Thread and Ethernet/Wi-Fi networks.

**Comment by dakhnod** [comment]
Thanks, this is highly informative and probably the way to go!

Still, I would be grateful to know if anything stop mdns from properly working.
Is it no possible to send multicast packets to ff02::fb?

**Comment by jwhui** [comment]
> Still, I would be grateful to know if anything stop mdns from properly working.
> Is it no possible to send multicast packets to ff02::fb?

mDNS was largely designed for networks where broadcast is relatively efficient (i.e. Ethernet). On constrained networks like Thread, there is not much capacity for multicast traffic.

**Comment by EskoDijk** [comment]
> Is it no possible to send multicast packets to ff02::fb?

Thread nodes can send link-local multicast to that address. It will only be received by direct neighboring nodes in the mesh that happen to listen to that address. As Jonathan mentioned, Thread has opted to use (unicast) SRP/DNS as a mechanism for mesh-wide and beyond service discovery, so by default no Thread node would respond to an mDNS query.

**Comment by EskoDijk** [comment]
@dakhnod @jwhui It seems like we can close this issue?
